src/services/completions.ts

```diff
@@ -162,27 +176,47 @@ namespace ts.Completions {
         log: Log,
         allowStringLiteral: boolean,
         symbolToOriginInfoMap?: SymbolOriginInfoMap,
-    ): Map<true> {
+    ): Map<{}> {
         const start = timestamp();
-        const uniqueNames = createMap<true>();
+        // Name -> map from module symbol id (or -1 for not from a module) to name
+        const uniques = createMap<true[]>();
         if (symbols) {
             for (const symbol of symbols) {
-                const entry = createCompletionEntry(symbol, location, performCharacterChecks, typeChecker, target, allowStringLiteral);
-                if (entry) {
-                    const id = entry.name;
-                    if (!uniqueNames.has(id)) {
-                        if (symbolToOriginInfoMap && symbolToOriginInfoMap[getUniqueSymbolId(symbol, typeChecker)]) {
-                            entry.hasAction = true;
-                        }
-                        entries.push(entry);
-                        uniqueNames.set(id, true);
-                    }
+                const origin = symbolToOriginInfoMap ? symbolToOriginInfoMap[getSymbolId(symbol)] : undefined;
+                const entry = createCompletionEntry(symbol, location, performCharacterChecks, typeChecker, target, allowStringLiteral, origin);
+                if (!entry) {
+                    continue;
+                }
+
+                const { name } = entry;
+                let uniquesForName = uniques.get(name);
+                if (!uniquesForName) {
+                    uniques.set(name, uniquesForName = []);

```

Comment:
this is gonna be sparse by definition. might be better to make it a map as well.

Reply:
we are guaranteed to see globals/locals before ones from modules, so:
- add a comment that explains that, since it is not obvious
- why not make it a map of number | [], or even split them into two maps global/local and modules, the reason is the global/local-only is by far the common case, and in that case we can avoid creating an extra array per entry.